### DISCUSSION TITLE
Multiples Routes to one function on API Gateway

### DISCUSSION THREAD

**Discussion by gwlester** [aws-powertools/powertools-lambda-python#537] [Q&A] [answered]
If I want multiple routes to a function, eg both **/hello** and **/v1/hello** to route to the same function can I:
`@app.get("/hello")
@app.get("/v1/hello")
def get_hello_universe():
    return {"message": "hello universe"}`

**Comment by heitorlessa** [comment, marked as answer]
**Updated**: We had further discussions on Slack about this which led to two features: [prefix stripping (/v1/hello, /v2/hello)](https://awslabs.github.io/aws-lambda-powertools-python/latest/core/event_handler/api_gateway/#custom-domain-api-mappings), and the Router for route splitting.

The prefix stripping feature addressed the main challenge: API Mappings. This features changes how API Gateway REST (v1) behaves when a request arrives. API Gateway adds the mapping as part of the route information within the event, leading to 404s and the original intent of wanting to register multiple routes for the same function.

Later and also related to this, we added support for [route splitting](https://awslabs.github.io/aws-lambda-powertools-python/latest/core/event_handler/api_gateway/#split-routes-with-router) to address the challenge of having routes defined in more than one file.

**Comment by ttopholm** [comment]
@heitorlessa was this ever implemented, because when I try, it fails

`

[ERROR] TypeError: 'NoneType' object is not callableTraceback (most recent call last):  File "/var/task/lambda_function.py", line 20, in lambda_handler    return app.resolve(event, context)  File "/var/task/aws_lambda_powertools/event_handler/api_gateway.py", line 498, in resolve    return self._resolve().build(self.current_event, self._cors)  File "/var/task/aws_lambda_powertools/event_handler/api_gateway.py", line 557, in _resolve    return self._call_route(route, match_results.groupdict())  # pass fn args  File "/var/task/aws_lambda_powertools/event_handler/api_gateway.py", line 611, in _call_route    return ResponseBuilder(self._to_response(route.func(**args)), route) | [ERROR] TypeError: 'NoneType' object is not callable Traceback (most recent call last):   File "/var/task/lambda_function.py", line 20, in lambda_handler     return app.resolve(event, context)   File "/var/task/aws_lambda_powertools/event_handler/api_gateway.py", line 498, in resolve     return self._resolve().build(self.current_event, self._cors)   File "/var/task/aws_lambda_powertools/event_handler/api_gateway.py", line 557, in _resolve     return self._call_route(route, match_results.groupdict()) # pass fn args   File "/var/task/aws_lambda_powertools/event_handler/api_gateway.py", line 611, in _call_route     return ResponseBuilder(self._to_response(route.func(**args)), route)
-- | --


`

Single routes works just fine

**Reply by heitorlessa** [reply]
Hey @ttopholm yep, we launched the Router to split routes into multiple files within a single Lambda function. If you refer to something else, we'd love a bug report to better understand the issue

https://awslabs.github.io/aws-lambda-powertools-python/latest/core/event_handler/api_gateway/#split-routes-with-router

**Reply by ttopholm** [reply]
Yes, but the question from @gwlester was too use multiple routes with one function, and that is failing.

Multiple files works fine with the include_router.

**Reply by heitorlessa** [reply]
Thanks for the correction - **no**, we don't and won't support this feature, as it can lead to a sub-optimal design. We had further discussions on Slack about this which led to two features: [prefix stripping (/v1/hello, /v2/hello)](https://awslabs.github.io/aws-lambda-powertools-python/latest/core/event_handler/api_gateway/#custom-domain-api-mappings), and the Router for route splitting.

I'll update my original answer to reflect that, as we didn't update the original question after the discussion too.

**Reply by heitorlessa** [reply]
I've just updated last year answer to include the original background (now lost on Slack), where this led to the prefix stripping feature when using API Mapping. We did not add support for multiple routes to the exact same function. What we did (after prefix stripping) was supporting distinct routes being defined in more than one file.

**Reply by ttopholm** [reply]
Thanks.
